### PR TITLE
Add Support for WebSockets

### PR DESCRIPTION
**Pull request by JarredAllen**
This PR adds support for WebSockets (as requested in #10).

This supports doing mqtt over an insecure websocket or over a secure websocket with default TLS configuration (I didn't implement custom TLS config for secure websockets because it would be more work and I don't need that functionality for my use cases, but it shouldn't be too hard to add if someone else wants it).

I did also have to move erroring on a failed read from happening immediately to only raising the error on an attempt to read an incoming packet, since the websocket implementation would otherwise raise an error when you disconnect. Hopefully that's acceptable.

### CODE REVIEW

**Comment by fluffysquirrels** [comment]
Thanks very much for this, I'll take a look at it in detail shortly.

**Comment by JarredAllen** [comment]
Are there any updates on this? I don't mean to pressure you (thanks for your work on this great library), but I'd greatly appreciate it if you could give an update on the status of this PR.

**Comment by fluffysquirrels** [comment]
Hi,

Sorry for the delay on this.

One problem I have is that in WebSockets mode the config is set differently than in TCP or TLS mode. In TCP or TLS mode `.host()` sets hostname and `.port()` sets port number, but in WebSockets mode `.host()` sets a URL and `.port()` is ignored. At the very least I think these should be consistent, but I think I prefer using URL for all modes, as this makes it simpler for a user to use their own config file and switch between protocols without caring about the details of what protocol is in use. This will require a breaking change, but the library is fairly young, and only a simple usage of `format!` should be required in the consumer. The [MQTT standard](https://github.com/mqtt/mqtt.org/wiki/URI-Scheme#scheme) suggests `mqtt` and `mqtts` protocols for the URL, which I like, and then we can use `ws` and `wss` for WebSockets URLs.

Similarly, your patch as it is supports setting TLS client config for TLS mode, but not for secure WebSockets mode (i.e. `wss`). I'd like to see `ClientBuilder.set_tls_client_config()` pass through the configuration to `tokio-tungstenite`. If your configuration is anything like mine, you're using self-signed certificates on your servers and setting TLS client config to securely connect to self-signed certificates requires custom TLS client config.

I have a few other comments that I'll make against the appropriate lines of code.

File: Cargo.toml [hunk 1]
```diff
@@ -12,19 +12,22 @@ repository = "https://github.com/fluffysquirrels/mqtt-async-client-rs"
 bytes = "0.4.0"
 futures-core = "0.3.1"
 futures-util = "0.3.1"
+http = "0.2.4"
```
Comment (fluffysquirrels) [inline Cargo.toml:15]: Can `http` be optional and included on the "websocket" feature?

File: Cargo.toml [hunk 2]
```diff
@@ -12,19 +12,22 @@ repository = "https://github.com/fluffysquirrels/mqtt-async-client-rs"
...
+tokio-tungstenite = { version = "0.14", optional = true, features = ["rustls-tls"] }
 
 [dev-dependencies]
 env_logger = "0.7.1"
 structopt = "0.3.5"
 webpki-roots = "0.18.0"
 
 [features]
-default = ["tls"]
+default = ["tls", "websocket"]
```
Comment (fluffysquirrels) [inline Cargo.toml:30]: I think we should not include the "websocket" feature by default.

Also, please document the "websocket" feature in `${REPO}/src/lib.rs` as the "tls" feature is.

File: src/client/client.rs [hunk 3]
```diff
@@ -546,26 +547,30 @@ impl Client {
...
-    }
-
-    #[cfg(not(feature = "tls"))]
-    {
-        let tcp = TcpStream::connect((&*opts.host, opts.port)).await?;
-        Ok(AsyncStream::TcpStream(tcp))
+        #[cfg(feature = "websocket")]
+        ConnectionMode::Websocket => {
+            // let url = format!("{}:{}", opts.host, opts.port);
+            // println!("Websocket: \"{}\"", url);
```
Comment (fluffysquirrels) [inline src/client/client.rs:567]: Can we either remove these 2 lines, or if they're till useful log using `debug!`?

File: src/client/client.rs [hunk 4]
```diff
@@ -1056,6 +1066,22 @@ impl IoType {
     }
 }
 
+/// An enum for specifying which mode we will use to connect to the broker
+#[derive(Clone)]
+pub enum ConnectionMode {
```
Comment (fluffysquirrels) [inline src/client/client.rs:1071]: This is a nice little enum!

File: tests/integration_test.rs [hunk 5]
```diff
@@ -180,6 +210,15 @@ fn tls_client() -> Result<Client> {
...
 
+#[cfg(feature = "websocket")]
+fn websocket_client() -> Result<Client> {
+    Client::builder()
+        .set_host("ws://127.0.0.1:9001".to_owned())
+        .set_websocket()
+        .set_connect_retry_delay(Duration::from_secs(1))
+        .build()
+}
+
```
Comment (fluffysquirrels) [inline tests/integration_test.rs:221]: I'd like to see a test that uses `wss` as well.

**Review by fluffysquirrels — CHANGES_REQUESTED** [review]

File: Cargo.toml (hunk 1 above)
Comment (JarredAllen) [reply Cargo.toml]: Good catch

File: src/client/client.rs (hunk 3 above)
Comment (JarredAllen) [reply src/client/client.rs:567]: Whoops. I thought I removed them, but I guess not.

File: Cargo.toml (hunk 2 above)
Comment (JarredAllen) [reply Cargo.toml]: Done!

**Comment by JarredAllen** [comment]
Okay, I've done those changes. I agree that it's nicer to use a url to set the remote host (I only didn't do that originally because I was hesitant to suggest a breaking change). I've also added in tls client configuration.

I am a bit hesitant on what port should be used when no port is specified. I left it just using 1883 by default, regardless of the connection type, since that's what it did before. However, I feel like it could be nicer if it defaulted to the port for the corresponding type (80 for ws, 443 for wss, 1883 for mqtt, 8883 for mqtts). If you agree that this would be better, I can do that change pretty easily.

File: tests/integration_test.rs (hunk 5 above)
Comment (JarredAllen) [reply tests/integration_test.rs]: Now added!

**Comment by fluffysquirrels** [comment]
> However, I feel like it could be nicer if it defaulted to the port for the corresponding type (80 for ws, 443 for wss, 1883 for mqtt, 8883 for mqtts).

I agree, please go ahead with this change.

**Comment by JarredAllen** [comment]
> > However, I feel like it could be nicer if it defaulted to the port for the corresponding type (80 for ws, 443 for wss, 1883 for mqtt, 8883 for mqtts).
> 
> I agree, please go ahead with this change.

Okay, done!

**Comment by fluffysquirrels** [comment]
@JarredAllen Just merged this into master with only a slight tweak to the `set_url` docs to mention the default port numbers.

Thanks again for your support and sorry for the delays in getting it merged. I'll be releasing v0.3.0 shortly with your changes.

**Pull request merged**